### ISSUE TITLE
Allow to have a coq.extraction and coq.theory in the same directory

### ISSUE THREAD

**Issue by lthms** [ocaml/dune#4158]
## Desired Behavior

To the best of my knowledge, it is currently not possible to have a `coq.theory` and
and `coq.extraction` within the same directory.

The reason is the following: `coq.extraction` uses the flag `-R . DuneExtraction` even
if this flag is useless (if we had a Coq file in `.`, `coq.extraction` will not know how
to build it). But in addition it prevents the use case I propose.

If I am right, this flag can be safely removed, and if we do, my use case should work.

## Example

See [this test repository to have a complete example](),

but in practice the `dune` file I am looking for to be able to write is

```lisp
(coq.theory
  (modules :standard \ extract)
  (name Foo))

(coq.extraction
  (prelude extract)
  (extracted_modules extract)
  (theories Foo))
```

**Comment by lthms** [comment]
Writing this, I wonder to which extend `coq.theory` needs to be separated, and on the contrary could be merged.

Having a field of the form `extracted_modules` to `coq.theory` would be more generic, and probably better, no?

**Comment by lthms** [comment]
@ejgallego what do you think the best course of action is? I personally find appealing to add a `extracted_modules` field to `coq.theory`, but if you prefer I can try to make it so there is no `-R . DuneExtraction` for the call of `coqc` due to `coq.extraction`.

**Comment by ejgallego** [comment]
@lthms thanks a lot for the bug report and the nice reproduction case; indeed that's a problem which we should solve; I wonder myself what is the best way:

- adding `extracted_modules` seems fine, @rgrinberg was quite careful to implement things properly so the code would be modular, only caveat is that you would have to add an association list, there is a lot more discussion in #2178 and #3299
- dropping `-R . DuneExtraction` may also be a good idea, I don't remember the exact details but it is likely that it is there just because an artifact of the internal `coq_rules` API ; I don't think anything would stop working if we would do this, right? It seems the .vo object is not even considered as a target.

I'll let some time in case Rudi would like to comment, but we may try 2 which should be easier.

**Comment by ejgallego** [comment]
For reference this is the error on @lthms example:
```
        coqc .extract.aux,extract.{glob,ml,mli,vo} (exit 1)
(cd _build/default && /home/egallego/.opam/4.07.1/bin/coqc -q -w -native-compiler-disabled -native-compiler ondemand -Q . Foo -R . DuneExtraction extract.v)
While loading initial state:
Warning: /home/egallego/tmp/dune-issue-4158/_build/default was previously
bound to Foo; it is remapped to DuneExtraction
[overriding-logical-loadpath,loadpath]
File "./extract.v", line 1, characters 17-20:
Error: Cannot find a physical path bound to logical path matching suffix
<> and prefix Foo.
```

**Comment by lthms** [comment]
Yes, **2.** is a first good step, and I think you are right as for why `-R` was set to begin with. At least I had this impression while looking at the code.

I still think **1.** is a good idea in the long run, because it feels more… natural I would say, but it’s not critical either.

**Comment by rgrinberg** [comment]
To give some background why things are this way:

My original motivation was to make extraction/theory somewhat analogous to dune's library/executable concepts. I assumed this was a good thing because it allowed one to write a .v file with private definitions that wouldn't be potentially exposed to other users.

I also found it weird to thing about extraction in the context of public/private theories. If a theory is public, should the .ml sources it extracts be installed as well? Making this automatic didn't seem like a good idea to me back then, and the current approach maintains maximum flexibility as to what is done with the extracted files.

**Comment by lthms** [comment]
Thanks for the feedback. It makes sense indeed. Dropping `-R DuneExtraction`, or at least providing a way to set another namespace, seems therefore a more reasonable approach for now. I will try to provide a patch that does this so that we can move forward, and decide if it is a good way to deal with this or not.

**Comment by 229c9cf0** [comment]
The current behavior also makes it impossible to have extraction stubs on the top level, with libraries in sub directories, e.g.:

```
/runExtract.v
/theories/
    /foo/
        [...]
    /bar/
        [...]
```

Instead, you're forced to put them into disjoint directory trees, e.g:

```
/dummy/
    /runExtract.v
/theories/
    /foo/
        [...]
    /bar/
        [...]
```

This requirement isn't documented and cost me a while to understand. (New to Dune, so I kept wondering what I was doing wrong, instead of concluding that there's a bug...)

I find this rather annoying and would very much prefer not having to do so. (In my case, the code is generated from a C program and I will have to add extra wrappers to deal with directory stuff.) If it isn't changed to remove this requirement, it would be great if the behavior could at the very least be documented.

**Comment by ejgallego** [comment]
Hi @229c9cf0 , I'm sorry indeed that the state of this is not optimal, and we've been struggling as of late to find the manpower to improve it; any help is very welcome, so indeed if you could prepare a doc patch I'd be super happy to review and merge it.

Also I'd be very happy to help with improving the dune mode a bit more, we have an upcoming Coq hackthon where I will present the current dune rules for Coq and hopefully get to fix some of these bugs.
